base/sparse/linalg.jl

```diff
@@ -102,6 +102,34 @@ function (*){TX,TvA,TiA}(X::StridedMatrix{TX}, A::SparseMatrixCSC{TvA,TiA})
     Y
 end
 
+function (*){TvA,TiA}(X::Diagonal, A::SparseMatrixCSC{TvA,TiA})
+    mX, nX = size(X)
+    if nX != A.m
+        throw(DimensionMismatch("second dimension of X, $nX, must match first dimension of A, $(A.m)"))
+    end
+    Ynzval = zeros(promote_type(eltype(X),TvA), length(A.nzval))
+    rowval = A.rowval
+    nzval = A.nzval
+    @inbounds for k=1:length(nzval)
+        Ynzval[k] = X[rowval[k],rowval[k]] * nzval[k]
+    end
+    SparseMatrixCSC(A.m, A.n, copy(A.colptr), copy(A.rowval), Ynzval)
+end
+
+function (*){TvA,TiA}(A::SparseMatrixCSC{TvA,TiA}, X::Diagonal)
+    mX, nX = size(X)
+    if mX != A.n
+        throw(DimensionMismatch("second dimension of A, $(A.n), must match first dimension of X, $mX"))
+    end
+    Ynzval = zeros(promote_type(eltype(X),TvA), length(A.nzval))
+    rowval = A.rowval
+    nzval = A.nzval
+    @inbounds for col = 1:A.n, k=A.colptr[col]:(A.colptr[col+1]-1)
+        Ynzval[k] += X[col, col] * nzval[k]
```

Comment:
This shouldn't need to be `+=` since `colptr` is nondecreasing. Also would be better to use `Array` rather than `zeros` to initialize the nonzeros since you can make sparse matrices with element types that don't have `zero` defined.